### ISSUE TITLE
@modelOptions() in SubDocument not correct work

### ISSUE THREAD

**Issue by Manfredik** [typegoose/typegoose#133]
## `@modelOptions({ schemaOptions: { _id: false } })` - this options does not work

If you use `@modelOptions ({schemaOptions: {_id: false}})` for a sub-document model, all added sub-documents still have _id

## Code Example
```ts
@modelOptions({ schemaOptions: { _id: false } }) // this model options not work
export class SubDocument{
	@prop({ type: Number, min: 1, max: 5 })
	id!: number;
	@prop({ type: Number, min: 0, max: 4 })
	position!: number;
}

@plugin(MongooseAutoIncrementID.plugin, { modelName: 'UserUnit', field: 'id' })
@modelOptions({ schemaOptions: { collection: 'user_unit' } }) // this model option work correctly
export class UserUnit {
	@prop({ type: Number })
	id!: number;
	@prop({ type: Number, required: true })
	userId!: number;
	@arrayProp({ items: SubDocument})
	subDocument!: Types.Array<SubDocument>;
}

const unit  = await UserUnitModel.findOne({id});
const id = 1;
const position = 2;
unit.items.addToSet({id, position});
await unit.save();

console.log(unit.items);
/**
 [
   {
     "_id": "5dda88124ccc7d420c88fab2",
     "id":  1,
     "position": 2
   }
]

*/
```

**Comment by hasezoey** [comment]
a02490240050e11649a1b9e7334ae5cb7a84cc61 should fix it, could you confirm?

**Comment by Manfredik** [comment]
Yes, but this is not the correct library, in the original version of mongoose when using the _id: false directive, the _id field is not added to the sub-documents.
This incorrect observation leads to the inability to use normally some methods to work with the array of sub-documents.

for example
``ts
unit.items.pull({id: 1, position: 1});
``
Use the following code instead
``ts
const f = unit.items.find(i => i.id === 1 && i.position == 1);
if (typeof f !== 'undefined') {
   unit.items.pull(f);
}
``
Agree that this is not the best way

**Comment by hasezoey** [comment]
@Manfredik i dont understand what you are trying to say, because `_id: false` is to disable `_id` generation in sub-documents / sub-document-arrays

**Comment by Manfredik** [comment]
I 'm sorry, I didn 't see your fix

**Comment by Manfredik** [comment]
how can i check your fix ?

**Comment by hasezoey** [comment]
> how can i check your fix ?

i really need to do an guide for this...

---

`npm i -s github:typegoose/typegoose` (or add `#branch` for a specific branch)
`yarn add github:typegoose/typegoose` (or add `#branch` for a specific branch)
(yarn is known to NOT execute the `prepare` script (building the project), you either need to manually execute it or use npm (or other "inconvenient" ways))

**Comment by Manfredik** [comment]
not work 


[2019-11-28T14:10:37.197] [ERROR] [TSED] - TypeError: Schema is required and must be an instance of Mongoose Schema
    at new MongooseAutoIncrementID (C:\Users\manfr\WebstormProjects\bat-server2\node_modules\mongoose-auto-increment-reworked\index.js:33:19)
    at plugin (C:\Users\manfr\WebstormProjects\bat-server2\node_modules\mongoose-auto-increment-reworked\index.js:198:16)
    at Schema.plugin (C:\Users\manfr\WebstormProjects\bat-server2\node_modules\@typegoose\typegoose\node_modules\mongoose\lib\schema.js:1400:3)
    at Object._buildSchema (C:\Users\manfr\WebstormProjects\bat-server2\node_modules\@typegoose\typegoose\src\internal\schema.ts:74:11)
    at buildSchema (C:\Users\manfr\WebstormProjects\bat-server2\node_modules\@typegoose\typegoose\src\typegoose.ts:168:9)
    at Object.getModelForClass (C:\Users\manfr\WebstormProjects\bat-server2\node_modules\@typegoose\typegoose\src\typegoose.ts:103:58)
    at Object.<anonymous> (C:\Users\manfr\WebstormProjects\bat-server2\src\models\Unit.ts:67:26)
    at Module._compile (internal/modules/cjs/loader.js:936:30)
    at Module.m._compile (C:\Users\manfr\WebstormProjects\bat-server2\node_modules\ts-node\src\index.ts:530:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:947:10)
    at Object.require.extensions.<computed> [as .ts] (C:\Users\manfr\WebstormProjects\bat-server2\node_modules\ts-node\src\index.ts:533:12)
    at Module.load (internal/modules/cjs/loader.js:790:32)
    at Function.Module._load (internal/modules/cjs/loader.js:703:12)
    at Module.require (internal/modules/cjs/loader.js:830:19)
    at require (internal/modules/cjs/helpers.js:68:18)
    at Object.<anonymous> (C:\Users\manfr\WebstormProjects\bat-server2\src\models\UserUnit.ts:3:1)
    at Module._compile (internal/modules/cjs/loader.js:936:30)
    at Module.m._compile (C:\Users\manfr\WebstormProjects\bat-server2\node_modules\ts-node\src\index.ts:530:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:947:10)
    at Object.require.extensions.<computed> [as .ts] (C:\Users\manfr\WebstormProjects\bat-server2\node_modules\ts-node\src\index.ts:533:12)
    at Module.load (internal/modules/cjs/loader.js:790:32)
    at Function.Module._load (internal/modules/cjs/loader.js:703:12)
[nodemon] app crashed - waiting for file changes before starting...

**Comment by hasezoey** [comment]
@Manfredik it seems like an plugin error - what did you use, npm or yarn?

**Comment by Manfredik** [comment]
After installing the package using yarn
I executed npm install from the typegoose directory after I executed npm run prepare

**Comment by hasezoey** [comment]
@Manfredik could you try 6.1.6(typegoose)? it includes the changes

**Comment by Manfredik** [comment]
Great, bug fixed, tnx

**Comment by nasr18** [comment]
@Manfredik which `auto-increment` plugin you are using? and how you are initializing?

@hasezoey how to initialize `mongoose-auto-increment`?

**Comment by hasezoey** [comment]
@nasr18 i never used it, but as in the issue description, try (with `mongoose-auto-increment`):
```ts
import * as mai from "mongoose-auto-increment"; // i dont actually know if it is such an import, its just a guess

@plugin(mai.plugin, { modelName: 'SomeClass', field: 'some_auto_id' })
class SomeClass {
  @prop()
  public  some_auto_id: number;
}
```

**Comment by nasr18** [comment]
@hasezoey it will throw error that, mai is not initialized. For initialization it needs mongoose connection. How do I pass it?

**Comment by hasezoey** [comment]
@nasr18 like i said i never used `mongoose-auto-increment` so i dont really know how to use it
-> but from what i get from the documentation, the steps are:
1. connect to database
2. initialize the plugin with `mai.initalize(mongoose.connection)`
3. create the models

**Comment by nasr18** [comment]
@hasezoey `mongoose-auto-increment` plugin will create one collection and use that collection to auto-increment the value. Looks like that collection is not at all created and it keep on trying to create/find that collection. Inside this plugin, there's an `setTimeout` function to try to connect to that collection. so no response/error while creating new record. please help me to overcome this.

**Comment by hasezoey** [comment]
@nasr18 sorry i cant help you further with that because this issue is already solved and closed **and** i dont know anything about `mongoose-auto-increment`

**Comment by nasr18** [comment]
@hasezoey I think it's looking for typegoose class instead of mongoose schema. I'll try tomorrow and update.

**Comment by nasr18** [comment]
@hasezoey can you please check this issue? exact issue I'm facing in our project.

https://github.com/kpfromer/nestjs-typegoose/issues/27#issuecomment-582460458

**Comment by hasezoey** [comment]
@nasr18 on your request, i took a deeper look into `mongoose-auto-increment` (and later on `mongoose-auto-increment-reworked`), an i discovered that both are archived and didnt have an update in at least ~2 years - but i still tested the reworked one (the older one would from design just not work) and i discovered, it sets itself up correctly, but just dosnt want to increment the set prop (the pre-save hook on it never gets called and i dont know why)

here are the steps i took:
```sh
npm i -s mongoose-auto-increment-reworked
```

here the code:
```ts
// NodeJS: 12.14.1
// MongoDB: 4.2-bionic (Docker)
import { getModelForClass, prop, plugin, buildSchema, addModelToTypegoose } from "@typegoose/typegoose"; // @typegoose/typegoose@6.2.2
import * as mongoose from "mongoose"; // mongoose@5.8.9
import { MongooseAutoIncrementID as mai } from "mongoose-auto-increment-reworked";
import { getName } from "@typegoose/typegoose/lib/internal/utils";

@plugin(mai.plugin, { modelName: getName(SomeClass), field: "autoincremented", incrementBy: 1, unique: false, startAt: 1 })
class SomeClass {
  @prop({ required: true })
  public normalProp!: number;

  @prop({ required: true })
  public autoincremented!: number;
}
const SomeModel = getModelForClass(SomeClass);
/* the commented-out code below is just another way of applying the plugin, which results in the same */
// const SomeSchema = buildSchema(SomeClass);
// SomeSchema.plugin(mai.plugin, { modelName: getName(SomeClass), field: "autoincremented", incrementBy: 1, unique: false });
// const SomeModel = addModelToTypegoose(mongoose.model(getName(SomeClass), SomeSchema), SomeClass);

(async () => {
  await mongoose.connect(`mongodb://localhost:27017/`, { useNewUrlParser: true, dbName: "verify133", useCreateIndex: true, useUnifiedTopology: true });

  const initalDoc = await SomeModel.create({ normalProp: 1, autoincremented: 1 });
  console.log(initalDoc);

  const foundDoc = await SomeModel.findById(initalDoc._id).exec();
  await foundDoc?.save(); // the "?" is because it could be undefined, but in this exact case it is not, so just ignore it
  console.log(foundDoc); // not incremented

  await mongoose.disconnect();
})();
```

**Comment by hasezoey** [comment]
Update to my last comment: the hook gets triggered, but for logging it calls `debug` instead of `log` - that is why i missed it, and otherways it uses the wrong values - and it uses the outdated `findOneAndUpdate`

and because it is archived, has some much overhead (at least code-wise), the logging is not working correctly and simply dosnt work, i will probably attempt to make such an plugin myself - i will write it here when its done / ready for testing

**Comment by nasr18** [comment]
@hasezoey thanks for your detailed reply. Actually I'm trying to re-write the `mongoose-auto-increment` using `nestjs-typegoose`. But I can't make it. Bcos in nestjs-typegoose, the way they are implementing the mongoose model is different than the plugin.

**Comment by nasr18** [comment]
`import { Logger } from '@nestjs/common';
import { InjectModel } from 'nestjs-typegoose';
import { IdentityCounter } from '../types/src/models/identity-counter.model';
import { ReturnModelType } from '@typegoose/typegoose';

export class IdentityCounterPlugin {
	// Default settings and plugin scope variables.
	settings = {
		model: null, // The model to configure the plugin for.
		field: '_id', // The field the plugin should track.
		startAt: 0, // The number the count should start at.
		incrementBy: 1, // The number by which to increment the count each time.
		unique: true, // Should we create a unique index for the field
	};
	fields = {}; // A hash of fields to add properties to in Mongoose.
	ready = false; // True if the counter collection has been updated and the document is ready to be saved.

	schema = {};
	options = {}

	constructor(
		schema,
		options,
		@InjectModel(IdentityCounter)
		private readonly idCounterModel?: ReturnModelType<typeof IdentityCounter>,
	) {
		Logger.log('constructing...', 'lib -> mongoosePlugin');
		if (!schema) throw new Error('Schema is required');

		this.schema = schema;
		this.options = options;
	}

	initialize() {
		Logger.log('initializing...', 'lib -> mongoosePlugin');
		this.idCounterModel
			.findOne()
			.lean()
			.then(counterDoc => {
				Logger.log(counterDoc, 'lib -> mongoosePlugin -> counterDoc');
				if (!counterDoc) {
					const { model, field, startAt, incrementBy } = this.settings;
					const counterData = {
						model,
						field,
						count: startAt - incrementBy,
					};

					this.idCounterModel.create(counterData).then(newCounterDoc => {
						Logger.log(newCounterDoc, 'lib -> mongoosePlugin -> newCounterDoc');
						this.ready = true;
					});
				} else this.ready = true;
			});
	}

	static plugin(schema, options) {
		Logger.log(options, 'lib -> mongoosePlugin -> plugin');
		if (!options) throw new Error('Please provide options');

		if (!options.model) throw new Error('Please provide model name in options');

		return new IdentityCounterPlugin(schema, options).applyPlugin();
	}

	validateOptions() {
		const typeOfOptions = typeof this.options;
		Logger.log(typeOfOptions, 'validating options');
		switch (typeOfOptions) {
			case 'string':
				this.settings.model = this.options;
				break;

			case 'object':
				Object.assign(this.settings, this.options);
				Logger.log(this.settings, 'lib -> mongoosePlugin -> validateOptions object');
				break;
		}
	}

	applyPlugin() {
		Logger.log('applying plugin...', 'lib -> mongoosePlugin -> applyPlugin');
		this.validateOptions();
		this.initialize();
	}
}
`
this is what I tried and getting `findOne` is undefined inside `initialize` method. @hasezoey

**Comment by hasezoey** [comment]
> Actually I'm trying to re-write the mongoose-auto-increment using nestjs-typegoose. But I can't make it. Bcos in nestjs-typegoose, the way they are implementing the mongoose model is different than the plugin.

i dont understand what you are trying to do, when you just want a hard-coded version without much overhead:
```ts
@pre<YourClass>("save", function () {
  if (!this.isNew) {
    // only apply the increment if it is not new (to let the default be the default and not default+whatever
    this.toincrementprop += 1;
  }
}
export class YourClass {}
```
^ Note: this is untested, but something *like this* should work

**Comment by hasezoey** [comment]
@nasr18 [this](https://github.com/typegoose/auto-increment) might be what you want, right? please note that this is just an prototype, and so not on npm *yet*

**Comment by nasr18** [comment]
@hasezoey I can't able to import your plugin. think its not in npm. just getting empty folder. anyway I'm using source code from github.

**Comment by hasezoey** [comment]
@nasr18 thanks, should have checked that one, but forgot

should be fixed with 0.1.2
